File: sdks/node/src/vectors.ts
import type { ApplAdServer } from './client';

export class Vectors {
  constructor(private client: ApplAdServer) {}

  createIndex(indexId: string, dimensions: number, opts?: { metric?: string; name?: string }) {
    return this.client.call('POST', '/vectors/indexes', { indexId, dimensions, ...opts });
  }

  listIndexes() {
    return this.client.call('GET', '/vectors/indexes');
  }

  getIndex(indexId: string) {
    return this.client.call('GET', `/vectors/indexes/${indexId}`);
  }

  deleteIndex(indexId: string) {
    return this.client.call('DELETE', `/vectors/indexes/${indexId}`);
  }

  upsert(indexId: string, vectors: { id: string; values: number[]; metadata?: Record<string, unknown> }[]) {
    return this.client.call('POST', `/vectors/indexes/${indexId}/vectors`, { vectors });
  }

  query(indexId: string, vector: number[], opts?: { topK?: number; filter?: Record<string, unknown> }) {
    return this.client.call('POST', `/vectors/indexes/${indexId}/query`, { vector, ...opts });
  }

  deleteVectors(indexId: string, ids: string[]) {
    return this.client.call('POST', `/vectors/indexes/${indexId}/delete`, { ids });
  }
}

File: sdks/dart/lib/src/vectors.dart
import 'package:dio/dio.dart';

class Vectors {
  final Dio _dio;

  Vectors(this._dio);

  Future<Map<String, dynamic>> createIndex({
    required String indexId,
    required int dimensions,
    String? metric,
    String? name,
  }) async {
    final res = await _dio.post('/v1/vectors/indexes', data: {
      'indexId': indexId,
      'dimensions': dimensions,
      if (metric != null) 'metric': metric,
      if (name != null) 'name': name,
    });
    return res.data;
  }

  Future<Map<String, dynamic>> listIndexes() async {
    final res = await _dio.get('/v1/vectors/indexes');
    return res.data;
  }

  Future<Map<String, dynamic>> getIndex(String indexId) async {
    final res = await _dio.get('/v1/vectors/indexes/$indexId');
    return res.data;
  }

  Future<void> deleteIndex(String indexId) async {
    await _dio.delete('/v1/vectors/indexes/$indexId');
  }

  Future<Map<String, dynamic>> upsert({
    required String indexId,
    required List<Map<String, dynamic>> vectors,
  }) async {
    final res = await _dio.post('/v1/vectors/indexes/$indexId/vectors', data: {
      'vectors': vectors,
    });
    return res.data;
  }

  Future<Map<String, dynamic>> query({
    required String indexId,
    required List<double> vector,
    int? topK,
    Map<String, dynamic>? filter,
  }) async {
    final res = await _dio.post('/v1/vectors/indexes/$indexId/query', data: {
      'vector': vector,
      if (topK != null) 'topK': topK,
      if (filter != null) 'filter': filter,
    });
    return res.data;
  }

  Future<Map<String, dynamic>> deleteVectors({
    required String indexId,
    required List<String> ids,
  }) async {
    final res = await _dio.post('/v1/vectors/indexes/$indexId/delete', data: {
      'ids': ids,
    });
    return res.data;
  }
}
